### PR TITLE
Generate: add test for left-padding support

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

This PR adds a test to check whether a decoder-only model supports left padding. The test was somewhat tricky to design -- the reasoning is all commented in the code, for future reference.

Let me know if you agree with the decisions made in the test!

Hopefully we will be able to detect whether a model supports left-padding before merging 🤞

### CODE REVIEW

**Comment by gante** [comment]
Current failures:
Screenshot 2023-03-22 at 18 42 57

File: tests/generation/test_utils.py [hunk 1]
```diff
@@ -1497,6 +1497,54 @@ def test_generate_with_head_masking(self):
                 attn_weights = out[attn_name] if attn_name == attention_names[0] else out[attn_name][-1]
                 self.assertEqual(sum([w.sum().item() for w in attn_weights]), 0.0)
 
+    # TODO (joao): this test is actually not slow :) However, it is not passing in some models (e.g. GPTNeoX) and the
+    # fix for some models is quite lengthy. Being slow means it doesn't block our push CI while we fix it.
+    @slow
+    def test_left_padding_compatibility(self):
+        # The check done in this test is fairly difficult -- depending on the model architecture, passing the right
+        # position index for the position embeddings can still result in a different output, due to numerical masking.
```
Comment (ydshieh) [inline tests/generation/test_utils.py:1505]: just curious, why we use the term `numerical` here and below? Is `attention mask` not enough?

File: tests/generation/test_utils.py [hunk 2]
```diff
@@ -1497,6 +1497,54 @@ def test_generate_with_head_masking(self):
...
+
+        for model_class in self.all_generative_model_classes:
+            config, _, _, _ = self._get_input_ids_and_config()
+            if config.is_encoder_decoder:
+                continue  # skip for encoder-decoder models -- they don't need left-padding compatibility
+            model = model_class(config).to(torch_device).eval()
+            signature = inspect.signature(model.forward).parameters.keys()
+
+            no_failures = True
+            for _ in range(30):
```
Comment (ydshieh) [inline tests/generation/test_utils.py:1523]: do you know how long this test takes in general (with `30` in the loop).

I would prefer not to use `30` - if `5` sounds good to you.

File: tests/generation/test_utils.py [hunk 3]
```diff
@@ -1497,6 +1497,54 @@ def test_generate_with_head_masking(self):
...
+                _, input_ids, attention_mask, _ = self._get_input_ids_and_config()
+                model_kwargs = {"input_ids": input_ids, "attention_mask": attention_mask}
+                if "position_ids" in signature:
+                    position_ids = torch.cumsum(attention_mask, dim=-1) - 1
+                    position_ids.masked_fill_(attention_mask == 0, 1)
+                    model_kwargs["position_ids"] = position_ids
+                next_logits_wo_padding = model(**model_kwargs).logits[:, -1, :]
+
+                pad_size = (input_ids.shape[0], 1)
+                padding = torch.ones(pad_size, dtype=input_ids.dtype, device=torch_device) * config.pad_token_id
```
Comment (ydshieh) [inline tests/generation/test_utils.py:1533]: I don't check thoroughly, but it seems `config.pad_token_id` could be `None`. At least, in `PretrainedConfig`, the default is

```python
self.pad_token_id = kwargs.pop("pad_token_id", None)
```

Comment (ydshieh) [reply tests/generation/test_utils.py:1533]: And the following place doesn't always guarantee `pad_token_id ` is set.

https://github.com/huggingface/transformers/blob/8eb19e8b5d84808eb1c85c90da6bd10ae7efaceb/tests/generation/test_utils.py#L93

**Review by ydshieh — APPROVED** [review]
A few nit questions/comments, but LGTM overall.

Thank you for taking the left padding, @gante ❤️

File: tests/generation/test_utils.py (hunk 1 above)
Comment (gante) [reply tests/generation/test_utils.py:1505]: I want to make it explicit that attention mask is not a binary operation, but rather a numerical operation (and thus constrained to the maximum negative value of the variable format of choice). Generation is quite sensible to masking, especially at lower resolution representations with limited dynamic range like `float16`.

File: tests/generation/test_utils.py (hunk 2 above)
Comment (sgugger) [reply tests/generation/test_utils.py:1523]: That doesn't answer the question of how long it takes ;-) It's okay if we detect a false positive with the test being flaky compared to having a very long test.

Comment (gante) [reply tests/generation/test_utils.py:1523]: @sgugger @ydshieh 
Ah, on my local machine it takes 13.51s to run this test on ALL models. So less than 1 second per decoder-only model :)

Comment (ydshieh) [reply tests/generation/test_utils.py:1523]: Sound good, but maybe let me try on our CI runner which is not as fast as your local machine :-)

Comment (sgugger) [reply tests/generation/test_utils.py:1523]: Can we still put a smaller number before merging? Like 10? The tests are executed very often so we will see flaky failures. There are also tools to repeat a given tests multiple times in pytest.

**Review by sgugger — APPROVED** [review]
Thanks for working on this!

**Comment by gante** [comment]
@sgugger reduced to 10 runs, merging when CI gets green :)
